agent/consul/acl_client.go

```diff
@@ -0,0 +1,106 @@
+package consul
+
+import (
+	"time"
+
+	"github.com/hashicorp/consul/acl"
+	"github.com/hashicorp/consul/agent/metadata"
+	"github.com/hashicorp/consul/agent/structs"
+	"github.com/hashicorp/consul/lib"
+	"github.com/hashicorp/serf/serf"
+)
+
+var clientACLCacheConfig *structs.ACLCachesConfig = &structs.ACLCachesConfig{
+	// The ACL cache configuration on client agents is more conservative than
+	// on the servers. It is assumed that individual client agents will have
+	// fewer distinct identities accessing the client than a server would
+	// and thus can put smaller limits on the amount of ACL caching done.
+	//
+	// Identities - number of identities/acl tokens that can be cached
+	Identities: 1024,
+	// Policies - number of unparsed ACL policies that can be cached
+	Policies: 128,
+	// ParsedPolicies - number of parsed ACL policies that can be cached
+	ParsedPolicies: 128,
+	// Authorizers - number of compiled multi-policy effective policies that can be cached
+	Authorizers: 256,
+}
+
+func (c *Client) UseLegacyACLs() bool {
+	return c.useNewACLs.IsSet() == false
+}
+
+func (c *Client) monitorACLMode() {
+	waitTime := aclModeCheckMinInterval
+	for {
+		canUpgrade := false
+		for _, member := range c.LANMembers() {
+			if valid, parts := metadata.IsConsulServer(member); valid && parts.Status == serf.StatusAlive {
+				if parts.ACLs != structs.ACLModeEnabled {
+					canUpgrade = false
+					break
+				} else {
+					canUpgrade = true
+				}
+			}
+		}
+
+		if canUpgrade {
+			c.logger.Printf("[DEBUG] acl: transition out of legacy ACL mode")
+			c.useNewACLs.Set(true)
+			lib.UpdateSerfTag(c.serf, "acls", string(structs.ACLModeEnabled))
+			return
+		}
+
+		select {
+		case <-c.shutdownCh:
+			return
+		case <-time.After(waitTime):
```

Comment:
The serf tags are pushed as member updates. This go routine just "monitors" the local knowledge about those tags of the other members so there is 0 server interaction. For that reason I didn't see a need for a jitter here.

I had thought about attempting to plumb through getting member update events and keeping track of the systems that are still not in the desired state and that might make sense as an improvement in the future as it would make the transition more responsive and require less CPU on a massive cluster. However given the time constraints that wasn't going to make it for the initial implementation.